### PR TITLE
Added discovery and graphing for JunOS (SRX) RPM probes

### PR DESCRIPTION
**Pull request by PelNet**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

Hi,

I've added discovery support for SRX RPM probes for supported devices running JunOS. I believe the changes are relatively self-explanatory, but I've added a couple of screenshots per your stated guidelines. Please note that I'm using a custom dark theme instead of the default, though the RPM part works like any other graph/element.

2020-02-23-193250_1828x454_scrot
2020-02-23-193313_982x782_scrot

Please let me know if I missed something or if you require further info/samples/etc. 

Cheers,

PelliX

### CODE REVIEW

File: includes/discovery/functions.inc.php [hunk 1]
```diff
@@ -449,6 +450,9 @@ function sensor_limit($class, $current)
         case 'snr':
         case 'frequency':
         case 'pressure':
+        case 'probes':
+            $limit = 5;
```
Comment (Jellyfrog) [inline includes/discovery/functions.inc.php]: This changes the limit for airflow, Snr, frequency and pressure also, I'm guessing this is not intended?

**Comment by PipoCanaja** [comment]
Hi @PelNet 
Could you name this sensor "Percent" instead of "Probes", so it can be used in a more standard way? 
Probes, by itself, does not imply "percent", not like relative humidity which is always measured as percents. 
Thanx

File: includes/discovery/functions.inc.php (hunk 1 above)
Comment (PelNet) [reply includes/discovery/functions.inc.php]: Yes, good point. Will do.

**Comment by PelNet** [comment]
> Hi @PelNet
> Could you name this sensor "Percent" instead of "Probes", so it can be used in a more standard way?
> Probes, by itself, does not imply "percent", not like relative humidity which is always measured as percents.
> Thanx

Fair enough; I'll be a little pressed for time, so it might be Wednesday before I get back to you.

Cheers,

PelliX

**Comment by PipoCanaja** [comment]
No problem here, we have all the time needed :)

File: includes/discovery/functions.inc.php (hunk 1 above)
Comment (PelNet) [reply includes/discovery/functions.inc.php]: Fixed in latest commit, thanks :+1: 

PelliX

**Comment by Rosiak** [comment]
Mentioned here:
https://community.librenms.org/t/juniper-rpm-support/501/

**Comment by PelNet** [comment]
Hi,

I believe the issues mentioned are resolved in the last commit. Should I create a new PR or keep things in this one?

Cheers,

PelliX

**Comment by TheCheezeIt** [comment]
When this gets merged, I will be testing it out for you guys.

Is there anything specific you guys want me to test other than just some probes?

File: doc/Developing/os/Health-Information.md [hunk 2]
```diff
@@ -37,6 +37,7 @@ the values we expect to see the data in:
 | temperature                     | C                           |
 | voltage                         | V                           |
 | waterflow                       | l/m                         |
+| probes                          | %                           |
```
Comment (laf) [inline doc/Developing/os/Health-Information.md]: Needs updating to Percent here.

**Comment by PipoCanaja** [comment]
> When this gets merged, I will be testing it out for you guys.

Hi @TheCheezeIt 
It would be even better to test it before it is merged :) that way your feedback could be used to improve, if necessary, the PR before merging it. 
You can do it with `./scripts/github-apply 11187`

**Comment by PelNet** [comment]
Hi,

I believe the issues mentioned are resolved in the last commit. Should I create a new PR or keep things in this one? I've also uploaded the requested test SNMP data from an SRX device which has RPM configured.

Cheers,

PelliX
[junos.snmprec.txt](https://github.com/librenms/librenms/files/4270236/junos.snmprec.txt)

**Comment by PipoCanaja** [comment]
Can you please add the test data directly to this PR ? create a new variant (junos_rpm.snmprec and junos_rpm.json for instance) so that they do not collide with existing test-data.

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-64-release-changelog-may-2020/12290/1

**Comment by jpieta** [comment]
Heads up the divisor for this sensor is 1 for an SRX running Junos 15 and 1000000 for an SRX running Junos 18. Sharing because I didn't find this documented anywhere, perhaps changed to be able to show "five nines" reliability.

**Comment by dorkmatt** [comment]
@jpieta does this mean the recent JunOS builds need some additional conditions for the divisor?

**Comment by jpieta** [comment]
@dorkmatt Yes. I added a user function to junos.yaml to use a divisor of 1,000,000 if the value was greater than 100 in order to accommodate both Junos variants. 

Also, the oid calculation in functions.php wasn't working properly because it was calculating the index length including the name of the test. I added the following to functions.php as a quick fix:

function getRPMoid($index)
{
    $indexarr = explode('.',$index);
    $len = strlen($indexarr[0]);

    switch($indexarr[2]) {
        case "currentTest":
            $endindex = 1;
            break;
        case "LastCompletedTest":
            $endindex = 2;
            break;
        case "allTests":
            $endindex = 4;
            break;
        default:
            $endindex = 4;
            break;
    }
    $calcstring = $indexarr[0] . "\t" . $indexarr[1];
    $calcbin = implode(".", unpack("c*", $calcstring));
    $oid = $len . '.' . $calcbin . '.' . $endindex;
    //$oid = '.1.3.6.1.4.1.2636.3.50.1.2.1.4.' . $len . '.' . $calcbin . '.' . $endindex;

    return $oid;
}

**Comment by dorkmatt** [comment]
@jpieta Would you be willing to submit your fixes? It should be possible to add a condition based on the version number for the updated calculations.

**Comment by jpieta** [comment]
@dorkmatt sure thing. Here's the condition fix in /includes/discovery/functions.inc.php. I'll submit a PR this weekend.

function getRPMvalue($value)
{
	if($value > 100) {
		$value = $value / 1000000;
	}
	return $value;
}

**Pull request merged**